### ISSUE TITLE
FmtV() doesn't handle weird utf

### ISSUE THREAD

**Issue by kjk** [sumatrapdfreader/sumatrapdf#1386]
Seen in https://github.com/sumatrapdfreader/sumatrapdf/issues/878

When a file name contained unusual Unicode character, FmtV() would fail when it was given as `%S` and generate some garbage when manually converted to utf8 and given as `%s`. This is because of `vsnprintf`. Maybe change to `std::vsnprintf` or finish up StrFormat and use that instead.

**Comment by GitHubRulesOK** [comment]
Or like the zip message suggest when opening/saving such a file the user does not use characters in a document name that cause downstream problems with .synctex .zip .cbx etc

**Comment by SumatraPeter** [comment]
> Or like the zip message suggest when opening/saving such a file the user does not use characters in a document name that cause downstream problems with .synctex .zip .cbx etc

Naah, better to fix it than to tell folks in this day and age that they can't use Unicode chars in filenames despite running Sumatra on a modern OS! 🙂

**Comment by GitHubRulesOK** [comment]
@SumatraPeter I agree that if easy to work around OS failings, it can be worth it.
My suggestion is to avoid known problem characters that will inevitably raise future issues
If you include **&format** in a filename it is accepted but since command line calls use it for appending commands it is a potential exploit waiting to happen

**Comment by SumatraPeter** [comment]
> I agree that if easy to work around OS failings, it can be worth it.

kjk will correct me if I'm mistaken, but FmtV() doesn't seem to be part of the Win32 API and so to my untrained eye it looks like an issue with the program's code rather than an OS failing as such?

> If you include &format in a filename it is accepted but since command line calls use it for appending commands it is a potential exploit waiting to happen

Not sure what you mean by "potential exploit"? After all, neither MS-DOS nor Windows in all these _decades_ has ever prohibited **&** in filenames despite the special meaning for command.com/cmd.exe, so it seems silly for Sumatra to do so. If this was such a huge issue as to lead to a "potential exploit", it seems so obvious that surely MS would never even have allowed it in the first place (or fixed it ages ago)?

**Comment by GitHubRulesOK** [comment]
https://duckduckgo.com/?q=exploit+filename+%26+command&t=ffnt&ia=web

**Comment by SumatraPeter** [comment]
That seems to point to PowerShell not sanitising .ps1 script filenames before execution? What does a failure on the part of the PS devs have to do with Sumatra? I must confess I still do not see the potential for exploit in Sumatra - after all it _opens_ PDFs etc. with **&** in the filename just fine? TBH logic dictates to me that in this case MS should fix the bug in PS instead of throwing the baby out with the bathwater and completely blocking the use of **&** in filenames...

**Comment by GitHubRulesOK** [comment]
It is not just & 
One of the reasons I support SumatraPDF is its simplicity to not run coded files (such as forms and others) however it is not 100% immune to all exploits especially if some external file library execution is inevitable.
Having been previously requested to deal with Virus/Trojan/Repurposing self writing files including PDF structure I am wary of extending some features.

**Comment by SumatraPeter** [comment]
Of course Sumatra cannot be immune to (or responsible for) exploits targeting any third-party code/external library that it uses. But you know what, I'm not going to indulge in idle speculation here about the program's susceptibility to exploits in this case. As far as I can see Sumatra is not executing anything, and even if it was, debarring certain common characters from filenames sounds ridiculous to me. What's next, Sumatra should refuse to open/save PDF, SMX, VBKM etc. files with such 'problem' characters in their filenames? Again, logic would dictate that if the characters pose such a big issue, Sumatra should sanitize the filenames if required while opening. Beyond this I'll leave it to @kjk to weigh in on whether there's any need to be concerned here.
